src/collapse/ghcprof.rs

```diff
@@ -0,0 +1,222 @@
+                break;
+            }
+            let l = String::from_utf8_lossy(&line);
+            let line = l.trim_end();
+            if line.is_empty() {
+                break;
+            } else {
+                self.on_line(line, &mut occurrences, &cols)?;
+            }
+        }
+
+        // Write the results...
+        occurrences.write_and_clear(writer)?;
+
+        // Reset the state...
+        self.current_cost = 0;
+        self.stack.clear();
+        Ok(())
+    }
+
+    /// Check for start line of a call graph.
+    fn is_applicable(&mut self, input: &str) -> Option<bool> {
+        let mut input = input.as_bytes();
+        let mut line = String::new();
+        loop {
+            line.clear();
+            if let Ok(n) = input.read_line(&mut line) {
+                if n == 0 {
+                    break;
+                }
+            } else {
+                return Some(false);
+            }
+
+            if line
+                .split_whitespace()
+                .take(START_LINE.len())
+                .eq(START_LINE.iter().cloned())
+            {
+                return Some(true);
+            }
+        }
+        None
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(opt: Options) -> Self {
+        Folder {
+            opt,
+            ..Default::default()
+        }
+    }
+}
+
+impl Folder {
+    // Handle call graph lines of the form:
+    //
+    // MAIN           MAIN ...
+    //  CAF           Options.Applicative.Builder ...
+    //   defaultPrefs Options.Applicative.Builder ...
+    //    idm         Options.Applicative.Builder ...
+    //    prefs       Options.Applicative.Builder ...
+    //   fullDesc     Options.Applicative.Builder ...
+    //   hidden       Options.Applicative.Builder ...
+    //   option       Options.Applicative.Builder ...
+    //    metavar     Options.Applicative.Builder ...
+    //  CAF           Options.Applicative.Builder.Internal ...
+    //   internal     Options.Applicative.Builder.Internal ...
+    //   noGlobal     Options.Applicative.Builder.Internal ...
+    //   optionMod    Options.Applicative.Builder.Internal ...
+
+    fn on_line(
+        &mut self,
+        line: &str,
+        occurrences: &mut Occurrences,
+        cols: &Cols,
+    ) -> io::Result<()> {
+        if let Some(indent_chars) = line.find(|c| c != ' ') {
+            let prev_len = self.stack.len();
+            let depth = indent_chars;
+
+            if depth < prev_len {
+                // If the line is not a child, pop stack to the stack before the new depth
+                self.stack.truncate(depth);
+            } else if depth != prev_len {
+                return invalid_data_error!("Skipped indentation level at line:\n{}", line);
+            }
+            // There can be non-ascii names so take care to char offset not byte offset
+            let string_range = |col_start: usize| {
+                line.chars()
+                    .skip(col_start)
```

Comment:
This assumes that GHC's "thing that produces aligned output" aligns by Unicode code point, but that's not _necessarily_ the case. It could be aligning by glyph or grapheme or grapheme cluster instead, in which case this would not give us the right result. Someone mentioned that GHC has a `-pj` flag to produce JSON output — maybe it's worth digging into whether we should just use that instead?

Reply:
Not to mention, this would make us resistant to if GHC changes this (seemingly human-facing) format.

Reply:
this is the code that decides the length of function/module names etc.

https://github.com/ghc/ghc/blob/master/rts/ProfilerReport.c#L40

not sure what the implications are..

Reply:
It looks like they're trying to essentially detect what "kind" of UTF-8 byte they're looking at (based on the [codepage layout](https://en.wikipedia.org/wiki/UTF-8#Codepage_layout)). However I don't _think_ that'll actually come the same thing as `chars()` for code points with, say, 4 bytes in them :eyes: Based on the name of that function though, it certainly feels like what they're _going_ for is the utf-8 encoded string length (which is what `.chars()` will give you), so you may be okay :+1: